pkg/canary/reader/reader.go

```diff
@@ -344,12 +344,19 @@ func (r *Reader) run() {
 
 		// Set a read timeout of 10x the interval we expect to see messages
 		// Ignore the error as it will get caught when we call ReadJSON
-		_ = r.conn.SetReadDeadline(time.Now().Add(10 * r.interval))
+		timeoutInterval := 10 * r.interval
+		_ = r.conn.SetReadDeadline(time.Now().Add(timeoutInterval))
 		// I assume this is a blocking call that either reads from the websocket connection
 		// or times out based on the above SetReadDeadline call.
 		err := unmarshal.ReadTailResponseJSON(tailResponse, r.conn)
 		if err != nil {
-			fmt.Fprintf(r.w, "error reading websocket, will retry in 10 seconds: %s\n", err)
+			reason := "error reading websocket"
+			if e, ok := err.(net.Error); ok && e.Timeout() {
+				reason = fmt.Sprintf("timeout tailing new logs (timeout period: %.2fs)", timeoutInterval.Seconds())
```

Comment:
[Q]: May be I failed to understand something here. Can you help me understand?

Wondering how does it can help catching `log_format` errors in promtail? (which was original source of error)

Basically I'm trying to understand, what happens when setting wrong format in promtail in this case. that won't even send logs to Loki? or canary's `unmarshal.ReadTailResponseJSON` timesout because it was expecting the correct format? (in that should we add some message saying something like `timesout because it was expecting different log format`)

Reply:
How can we distinguish between a log line that doesn't match a filter and a log line that doesn't exist?

Reply:
@kavirajk 

> Basically I'm trying to understand, what happens when setting wrong format in promtail in this case. that won't even send logs to Loki? 

From my understanding, promtail does send logs to loki since you can actually run logcli to retrieve written logs.
E.g.
```
2021-11-25T14:20:07+01:00 {container="loki-canary", job="loki-dev-002/loki-canary", name="loki-canary", pod="loki-canary-8c49fcf8f-dm9qd", pod_template_hash="8c49fcf8f"}                                                       {"log":"1637846407189390762 ppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppp\n","stream":"stdout","time":"2021-11-25T13:20:07.189497945Z"}
```

> or canary's unmarshal.ReadTailResponseJSON timesout because it was expecting the correct format?

Not quite; the problem, I believe, is that the query that the canary reader sends to the ingesters does not match any log because of the format is different from the expected one. Hence, since there are no matching logs, the tailing operation times out since no logs are streamed to the canary reader within the given amount of time specified as the deadline.

Reply:
So the error is due to missing log line correct?

in that case, should we also error something on the promtail side? why logs not being sent?

Forgive my ignore there, basically I’m trying to relate how could we improve our error handling so that operator can easily catch the log format error in future!.

Reply:
Thanks @salvacorts for the clarification!. 

>I believe, is that the query that the canary reader sends to the ingesters does not match any log because of the format is different from the expected one

Can we somehow catch this behaviour?

(btw. not a blocker anymore!. this error handling makes sense for the actual error)

Reply:
I don't _think_ we can, unfortunately.

The log message will be pretty clear now though: it will timeout looking for the logs, so naturally the operator will try to validate that the logs are indeed being sent to Loki. They will most likely see that they are in the wrong format at that point.